test/test_op.py

```diff
@@ -0,0 +1,168 @@
+@pytest.mark.parametrize("order", [2, 3])
+@pytest.mark.parametrize(("vectorize", "nested"), [
+    (False, False),
+    (True, False),
+    (True, True)
+    ])
+def test_gradient(actx_factory, dim, form, order, vectorize, nested):
+    actx = actx_factory()
+
+    from pytools.convergence import EOCRecorder
+    eoc_rec = EOCRecorder()
+
+    for n in [4, 6, 8]:
+        mesh = mgen.generate_regular_rect_mesh(
+                a=(-1,)*dim, b=(1,)*dim,
+                nelements_per_axis=(n,)*dim)
+
+        dcoll = DiscretizationCollection(actx, mesh, order=order)
+
+        def f(x):
+            result = dcoll.zeros(actx) + 1.
+            for i in range(dim-1):
+                result *= actx.np.sin(np.pi*x[i])
+            result *= actx.np.cos(np.pi/2*x[dim-1])
+            return result
+
+        def grad_f(x):
+            result = make_obj_array([dcoll.zeros(actx) + 1. for _ in range(dim)])
+            for i in range(dim-1):
+                for j in range(i):
+                    result[i] *= actx.np.sin(np.pi*x[j])
+                result[i] *= np.pi*actx.np.cos(np.pi*x[i])
+                for j in range(i+1, dim-1):
+                    result[i] *= actx.np.sin(np.pi*x[j])
+                result[i] *= actx.np.cos(np.pi/2*x[dim-1])
+            for j in range(dim-1):
+                result[dim-1] *= actx.np.sin(np.pi*x[j])
+            result[dim-1] *= -np.pi/2*actx.np.sin(np.pi/2*x[dim-1])
+            return result
+
+        x = thaw(actx, op.nodes(dcoll))
+
+        if vectorize:
+            u = make_obj_array([(idim+1)*f(x) for idim in range(dim)])
+        else:
+            u = f(x)
+
+        def get_flux(u_tpair):
+            dd = u_tpair.dd
+            dd_allfaces = dd.with_dtag("all_faces")
+            normal = thaw(actx, op.normal(dcoll, dd))
+            u_avg = u_tpair.avg
+            if vectorize:
+                if nested:
+                    flux = make_obj_array([u_avg_i * normal for u_avg_i in u_avg])
+                else:
+                    flux = np.outer(u_avg, normal)
+            else:
+                flux = u_avg * normal
+            return op.project(dcoll, dd, dd_allfaces, flux)
+
+        dd_intfaces = DOFDesc("int_faces")
+        dd_allfaces = DOFDesc("all_faces")
+
+        if form == "strong":
+            # FIXME: this doesn't work
+            u_intfaces = op.project(dcoll, "vol", dd_intfaces, u)
+            grad_u = op.inverse_mass(dcoll,
+                op.local_grad(dcoll, u, nested=nested)
+                -  # noqa: W504
+                op.face_mass(dcoll,
+                    dd_allfaces,
+                    # Note: no boundary flux terms here because u_ext == u_int == 0
+                    get_flux(sym.TracePair(dd_intfaces,
+                        interior=u_intfaces,
+                        exterior=u_intfaces))
+                    -  # noqa: W504
+                    get_flux(op.interior_trace_pair(dcoll, u)))
+                )
```

Comment:
```suggestion
                - op.local_grad(dcoll, u, nested=nested)
                +  # noqa: W504
                op.face_mass(dcoll,
                    dd_allfaces,
                    # Note: no boundary flux terms here because u_ext == u_int == 0
                    get_flux(sym.TracePair(dd_intfaces,
                        interior=u_intfaces,
                        exterior=u_intfaces))
                    -  # noqa: W504
                    get_flux(op.interior_trace_pair(dcoll, u)))
                )
```
I think there is a sign error here?

Reply:
Huh. I noticed you flipped the signs consistently (even in the weak case). Nevermind.